test/test_selectors.py

```diff
@@ -0,0 +1,622 @@
+        self.addCleanup(signal.alarm, 0)
+
+        # Start the timer for the interrupt.
+        signal.alarm(1)
+
+        t = monotonic()
+        self.assertEqual([rd], selectors.wait_for_read(rd, timeout=2.0))
+        self.assertLess(monotonic() - t, 2.2)
+        self.assertEqual(rd.recv(1), b'x')
+
+
+@skipUnless(selectors.HAS_SELECT, "Platform doesn't have a selector and socketpair")
+class BaseSelectorTestCase(unittest.TestCase):
+    """ Implements the tests that each type of selector must pass. """
+    SELECTOR = selectors.DefaultSelector
+
+    def make_socketpair(self):
+        rd, wr = socket.socketpair()
+
+        # Make non-blocking so we get errors if the
+        # sockets are interacted with but not ready.
+        rd.settimeout(0.0)
+        wr.settimeout(0.0)
+
+        self.addCleanup(rd.close)
+        self.addCleanup(wr.close)
+        return rd, wr
+
+    def make_selector(self):
+        s = self.SELECTOR()
+        self.addCleanup(s.close)
+        return s
+
+    def standard_setup(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        s.register(rd, selectors.EVENT_READ, "data_rd")
+        s.register(wr, selectors.EVENT_WRITE, "data_wr")
+        return s, rd, wr
+
+    def test_get_key(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+
+        key = s.register(rd, selectors.EVENT_READ, "data")
+        self.assertEqual(key, s.get_key(rd))
+
+        # Unknown fileobj
+        self.assertRaises(KeyError, s.get_key, 999999)
+
+    def test_get_map(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+
+        keys = s.get_map()
+        self.assertFalse(keys)
+        self.assertEqual(len(keys), 0)
+        self.assertEqual(list(keys), [])
+        key = s.register(rd, selectors.EVENT_READ, "data")
+        self.assertIn(rd, keys)
+        self.assertEqual(key, keys[rd])
+        self.assertEqual(len(keys), 1)
+        self.assertEqual(list(keys), [rd.fileno()])
+        self.assertEqual(list(keys.values()), [key])
+
+        # Unknown fileobj
+        self.assertRaises(KeyError, keys.__getitem__, 999999)
+
+        # Read-only mapping
+        with self.assertRaises(TypeError):
+            del keys[rd]
+
+    def test_register(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+
+        # Ensure that the file is not yet added.
+        self.assertEqual(0, len(s.get_map()))
+        self.assertRaises(KeyError, lambda: s.get_map()[rd.fileno()])
+        self.assertRaises(KeyError, s.get_key, rd)
+        self.assertEqual(None, s._key_from_fd(rd.fileno()))
+
+        key = s.register(rd, selectors.EVENT_READ, "data")
+        self.assertIsInstance(key, selectors.SelectorKey)
+        self.assertEqual(key.fileobj, rd)
+        self.assertEqual(key.fd, rd.fileno())
+        self.assertEqual(key.events, selectors.EVENT_READ)
+        self.assertEqual(key.data, "data")
```

Comment:
We probably want to pass something we can assert identity on, rather than asserting equality.